### PR TITLE
Address issue #125307

### LINKED ISSUE #125307: Symbolic shapes unable to reason: Ne(Mod(u0*u2 + u1*u2, u0 + u1), 0)
### 🐛 Describe the bug

Repro:

```
import torch
from torch import Tensor
from typing import Optional
import torch._prims_common as utils
import torch._dynamo
        
torch._dynamo.config.capture_scalar_outputs = True


@torch.compile(fullgraph=True)
def f(x):
    a, b, stride = x.tolist()
    torch._check_is_size(a)
    torch._check_is_size(b)
    torch._check_is_size(stride)
    ta = torch.randn(a * stride)
    tb = torch.randn(b * stride)
    r = torch.cat([ta, tb])
    return r.view(a + b, stride)


x = torch.tensor([30, 20, 10])
f(x)
```

Fails with

```
torch._dynamo.exc.UserError: Tried to use data-dependent value in the subsequent computation. This can happen when we encounter unbounded dynamic value that is unknown during tracing time.  You will need to explicitly give hint to the compiler. Please take a look at constrain_as_value OR constrain_as_size APIs.  Could not guard on data-dependent expression Ne(Mod(u0*u2 + u1*u2, u0 + u1), 0) (unhinted: Ne(Mod(u0*u2 + u1*u2, u0 + u1), 0)).  (Size-like symbols: u2, u1, u0)

Potential framework code culprit (scroll up for full backtrace):
  File "/data/users/ezyang/b/pytorch/torch/_refs/__init__.py", line 3691, in _reshape_view_helper
    while guard_size_oblivious(accum % length != 0):

For more information, run with TORCH_LOGS="dynamic"
```

This is obviously false though, just need to be a bit smarter in reasoning.

cc @msaroufim @bdhirsh @anijain2305 @chauhang @lezcano 

### Versions

main

### PR DESCRIPTION
**Pull request by shazqadeer**
PyTorch overrides SymPy's Mod and does its own symbolic simplification. Inspired by issue #125307, this PR adds one more simplification tactic.

Fixes #125307 

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by shazqadeer** [comment]
@ezyang : I guessed at the syntax for the test. Please take a look.

File: test/test_proxy_tensor.py [hunk 1]
```diff
@@ -1108,6 +1108,20 @@ def forward(self, y_1, x_1):
...
+            torch._check_is_size(a)
+            torch._check_is_size(b)
+            torch._check_is_size(stride)
+            ta = torch.randn(a * stride)
+            tb = torch.randn(b * stride)
+            r = torch.cat([ta, tb])
+            return r.view(a + b, stride)
+
+        r = str(make_fx(f, tracing_mode="symbolic")(torch.tensor([30, 20, 10])).code).strip()
+        self.assertExpectedInline(r, "")
```
Comment (ezyang) [inline test/test_proxy_tensor.py:1123]: You'll need to update this so that it has the actual new output. But o/w it looks like it should work.

**Comment by ezyang** [comment]
Oh, tolist() doesn't work in make_fx. You can either call item() on things manually, or you can move the test to test/dynamo/test_misc.py or something (and use torch.compile to exercise it). Sorry, I had forgotten the exact contents of the test in the issue haha.

**Comment by shazqadeer** [comment]
```
import torch
from torch import Tensor
from typing import Optional
import torch._prims_common as utils
import torch._dynamo
from torch.fx.experimental.proxy_tensor import make_fx

torch._dynamo.config.capture_scalar_outputs = True


@torch.compile(fullgraph=True)
def f(_a, _b, _stride):
    a = _a.item()
    b = _b.item()
    stride = _stride.item()
    torch._check_is_size(a)
    torch._check_is_size(b)
    torch._check_is_size(stride)
    ta = torch.randn(a * stride)
    tb = torch.randn(b * stride)
    r = torch.cat([ta, tb])
    return r.view(a + b, stride)

_a = torch.tensor(30)
_b = torch.tensor(20)
_stride = torch.tensor(10)
f(_a, _b, _stride)

r = str(make_fx(f, tracing_mode="symbolic")(_a, _b, _stride).code).strip()
print(r)
```

I tried running the code above and got the following output:
```
(/home/shaz/local/a/pytorch-env) [shaz@devgpu005.nha1 ~/local/a/pytorch (issue-125307)]$ python ~/Temp/foo.py 
Traceback (most recent call last):
  File "/home/shaz/Temp/foo.py", line 29, in <module>
    r = str(make_fx(f, tracing_mode="symbolic")(_a, _b, _stride).code).strip()
  File "/data/users/shaz/a/pytorch/torch/fx/experimental/proxy_tensor.py", line 1403, in wrapped
    return make_fx_tracer.trace(f, *args)
  File "/data/users/shaz/a/pytorch/torch/fx/experimental/proxy_tensor.py", line 1349, in trace
    return self._trace_inner(f, *args)
  File "/data/users/shaz/a/pytorch/torch/fx/experimental/proxy_tensor.py", line 1336, in _trace_inner
    t = dispatch_trace(
  File "/data/users/shaz/a/pytorch/torch/_compile.py", line 24, in inner
    return torch._dynamo.disable(fn, recursive)(*args, **kwargs)
  File "/data/users/shaz/a/pytorch/torch/_dynamo/eval_frame.py", line 548, in _fn
    return fn(*args, **kwargs)
  File "/data/users/shaz/a/pytorch/torch/fx/experimental/proxy_tensor.py", line 641, in dispatch_trace
    graph = tracer.trace(root, concrete_args)
  File "/data/users/shaz/a/pytorch/torch/_dynamo/eval_frame.py", line 548, in _fn
    return fn(*args, **kwargs)
  File "/data/users/shaz/a/pytorch/torch/fx/_symbolic_trace.py", line 821, in trace
    (self.create_arg(fn(*args)),),
  File "/data/users/shaz/a/pytorch/torch/fx/experimental/proxy_tensor.py", line 659, in wrapped
    out = f(*tensors)
  File "/data/users/shaz/a/pytorch/torch/_dynamo/eval_frame.py", line 391, in _fn
    raise RuntimeError(
RuntimeError: Detected that you are using FX to symbolically trace a dynamo-optimized function. This is not supported at the moment.
```

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
From @shazqadeer 

```
Here is my code:

gcd = sympy.gcd(p, q)
        if gcd == q:
            print (p)
            print (q)
            print (gcd)
            return [S.Zero](http://s.zero/)

And here is the output:

s0/s1                                                                                                                                                                     
s0                                   
s0
```

**Comment by shazqadeer** [comment]
@ezyang : Some tests are failing but they seem unrelated to this PR. Ok to land?

**Review by lezcano — COMMENTED** [review]
I see the original reason to implement our own Mod is because of https://github.com/sympy/sympy/issues/25146, but that was already fixed. Can we use a newer SymPy version and completely remove this function?

**Comment by shazqadeer** [comment]
I think we should go ahead and land this PR. I will take an action item to investigate whether PyTorch's implementation of Mod can be dropped altogether given the current state of sympy. We can do that as a separate PR if things look promising.

@ezyang : What do you think?

**Comment by ezyang** [comment]
@lezcano well, we have to add our own version anyway to distinguish CMod and PythonMod, as well as Mod but only on non-negative arguments. So yes, I'd land this as is.

**Comment by wang-TJ-20** [comment]
> ```
> import torch
> from torch import Tensor
> from typing import Optional
> import torch._prims_common as utils
> import torch._dynamo
> from torch.fx.experimental.proxy_tensor import make_fx
> 
> torch._dynamo.config.capture_scalar_outputs = True
> 
> 
> @torch.compile(fullgraph=True)
> def f(_a, _b, _stride):
>     a = _a.item()
>     b = _b.item()
>     stride = _stride.item()
>     torch._check_is_size(a)
>     torch._check_is_size(b)
>     torch._check_is_size(stride)
>     ta = torch.randn(a * stride)
>     tb = torch.randn(b * stride)
>     r = torch.cat([ta, tb])
>     return r.view(a + b, stride)
> 
> _a = torch.tensor(30)
> _b = torch.tensor(20)
> _stride = torch.tensor(10)
> f(_a, _b, _stride)
> 
> r = str(make_fx(f, tracing_mode="symbolic")(_a, _b, _stride).code).strip()
> print(r)
> ```
> 
> I tried running the code above and got the following output:
> 
> ```
> (/home/shaz/local/a/pytorch-env) [shaz@devgpu005.nha1 ~/local/a/pytorch (issue-125307)]$ python ~/Temp/foo.py 
> Traceback (most recent call last):
>   File "/home/shaz/Temp/foo.py", line 29, in <module>
>     r = str(make_fx(f, tracing_mode="symbolic")(_a, _b, _stride).code).strip()
>   File "/data/users/shaz/a/pytorch/torch/fx/experimental/proxy_tensor.py", line 1403, in wrapped
>     return make_fx_tracer.trace(f, *args)
>   File "/data/users/shaz/a/pytorch/torch/fx/experimental/proxy_tensor.py", line 1349, in trace
>     return self._trace_inner(f, *args)
>   File "/data/users/shaz/a/pytorch/torch/fx/experimental/proxy_tensor.py", line 1336, in _trace_inner
>     t = dispatch_trace(
>   File "/data/users/shaz/a/pytorch/torch/_compile.py", line 24, in inner
>     return torch._dynamo.disable(fn, recursive)(*args, **kwargs)
>   File "/data/users/shaz/a/pytorch/torch/_dynamo/eval_frame.py", line 548, in _fn
>     return fn(*args, **kwargs)
>   File "/data/users/shaz/a/pytorch/torch/fx/experimental/proxy_tensor.py", line 641, in dispatch_trace
>     graph = tracer.trace(root, concrete_args)
>   File "/data/users/shaz/a/pytorch/torch/_dynamo/eval_frame.py", line 548, in _fn
>     return fn(*args, **kwargs)
>   File "/data/users/shaz/a/pytorch/torch/fx/_symbolic_trace.py", line 821, in trace
>     (self.create_arg(fn(*args)),),
>   File "/data/users/shaz/a/pytorch/torch/fx/experimental/proxy_tensor.py", line 659, in wrapped
>     out = f(*tensors)
>   File "/data/users/shaz/a/pytorch/torch/_dynamo/eval_frame.py", line 391, in _fn
>     raise RuntimeError(
> RuntimeError: Detected that you are using FX to symbolically trace a dynamo-optimized function. This is not supported at the moment.
> ```
> 
> So even using .item directly does not seem to help.
> 
> Should I move the test to test_misc.py?

hi，Has this problem been resolved

**Pull request closed without merging**